pkg/store/proxy.go

```diff
@@ -50,13 +50,127 @@ type Client interface {
 	component      component.StoreAPI
 	selectorLabels labels.Labels
 
 	responseTimeout time.Duration
 	metrics         *proxyStoreMetrics
 }
 
+func (s *ProxyStore) Rules(req *storepb.RulesRequest, srv storepb.Rules_RulesServer) error {
+	var (
+		g, gctx  = errgroup.WithContext(srv.Context())
+		respChan = make(chan *storepb.RuleGroup, 10)
+		groups   []*storepb.RuleGroup
+	)
+
+	defer func() { close(respChan) }()
+
+	for _, rulesClient := range s.rules() {
+		rs := &rulesStream{
+			client:  rulesClient,
+			request: req,
+			channel: respChan,
+			server:  srv,
+		}
+		g.Go(func() error { return rs.receive(gctx) })
+	}
+
+	for resp := range respChan {
+		groups = append(groups, resp)
+	}
+
+	if err := g.Wait(); err != nil {
+		level.Error(s.logger).Log("err", err)
+		return err
+	}
+
+	groups = dedupGroups(groups)
+
+	for _, g := range groups {
+		if err := srv.Send(storepb.NewRuleGroupRulesResponse(g)); err != nil {
+			return status.Error(codes.Unknown, errors.Wrap(err, "send rules response").Error())
+		}
+	}
+
+	return nil
+}
+
+func dedupGroups(groups []*storepb.RuleGroup) []*storepb.RuleGroup {
+	// sort groups such that they appear next to each other.
+	sort.Slice(groups, func(i, j int) bool { return groups[i].Name < groups[j].Name })
+
+	// nothing to do if we have no or a single result, also no deduplication is necessary.
+	if len(groups) < 2 {
+		return groups
+	}
+
+	i := 0
+	for _, g := range groups[1:] {
+		if g.Name == groups[i].Name {
+			groups[i].Rules = append(groups[i].Rules, g.Rules...)
+		} else {
+			i++
+			groups[i] = g
+		}
+	}
+
+	return groups[:i+1]
+}
+
+type rulesStream struct {
+	client  storepb.RulesClient
+	request *storepb.RulesRequest
+	channel chan *storepb.RuleGroup
+	server  storepb.Rules_RulesServer
+}
+
+func (stream *rulesStream) receive(ctx context.Context) error {
+	rules, err := stream.client.Rules(ctx, stream.request)
+	if err != nil {
+		err = errors.Wrapf(err, "fetching rules from rules client %v", stream.client)
+
+		if stream.request.PartialResponseStrategy == storepb.PartialResponseStrategy_ABORT {
+			return err
+		}
+
+		if err := stream.server.Send(storepb.NewWarningRulesResponse(err)); err != nil {
+			return err
+		}
+	}
+
+	for {
+		rule, err := rules.Recv()
+		if err == io.EOF {
+			return nil
+		}
+
+		if err != nil {
+			err = errors.Wrapf(err, "receiving rules from rules client %v", stream.client)
+
+			if stream.request.PartialResponseStrategy == storepb.PartialResponseStrategy_ABORT {
+				return err
+			}
+
+			if err := stream.server.Send(storepb.NewWarningRulesResponse(err)); err != nil {
+				return errors.Wrapf(err, "sending rules error to server %v", stream.server)
+			}
+
+			continue
+		}
+
+		if w := rule.GetWarning(); w != "" {
+			if err := stream.server.Send(storepb.NewWarningRulesResponse(errors.New(w))); err != nil {
+				return errors.Wrapf(err, "sending rules warning to server %v", stream.server)
+			}
+			continue
+		}
+
+		stream.channel <- rule.GetGroup()
```

Comment:
Don't you need to check here if the context has ended? There might not be any readers left

Reply:
indeed, great catch :+1: